src/app.js

```diff
@@ -1,59 +1,121 @@
+  var appSubs = props.subscriptions || []
   var appRoot = props.root || document.body
-  var element
-  var oldNode
+  var element = appRoot.children[0]
+  var beforeActionHooks = []
   var renderLock
 
-  appMixins.concat(props).map(function(mixin) {
-    mixin = typeof mixin === "function" ? mixin(emit) : mixin
+  initialize(appActions, props.actions, [])
 
-    Object.keys(mixin.events || {}).map(function(key) {
-      appEvents[key] = (appEvents[key] || []).concat(mixin.events[key])
-    })
-
-    appState = merge(appState, mixin.state)
-    initialize(appActions, mixin.actions)
+  appSubs.map(function(sub) {
+    if (typeof (sub = sub(appState, appActions)) === "function") {
+      beforeActionHooks.push(sub)
+    }
   })
 
-  requestRender(
-    (oldNode = emit("load", (element = appRoot.children[0]))) === element &&
-      (oldNode = element = null)
-  )
+  requestRender()
 
-  return emit
+  return appActions
 
-  function initialize(actions, withActions, lastName) {
+  function initialize(actions, withActions, lastPath) {
     Object.keys(withActions || {}).map(function(key) {
       var action = withActions[key]
-      var name = lastName ? lastName + "." + key : key
+      var path = lastPath.concat(key)
 
       if (typeof action === "function") {
         actions[key] = function(data) {
-          emit("action", { name: name, data: data })
-
-          var result = emit("resolve", action(appState, appActions, data))
-
-          return typeof result === "function" ? result(update) : update(result)
+          var afterActionHooks = []
+
+          beforeActionHooks.map(function(cb) {
+            if (
+              typeof (cb = cb({ name: path.join("."), data: data })) ===
+              "function"
+            ) {
+              afterActionHooks.push(cb)
+            }
+          })
+
+          var result = action(
+            getPath(lastPath, appState),
+            getPath(lastPath, appActions),
+            data
+          )
+
+          afterActionHooks.map(function(cb) {
+            result = cb(result)
+          })
+
+          return typeof result === "function"
+            ? result(function(withState) {
+                return update(lastPath, withState)
+              })
+            : update(lastPath, result)
         }
       } else {
-        initialize(actions[key] || (actions[key] = {}), action, name)
+        initialize(actions[key] || (actions[key] = {}), action, path)
       }
     })
   }
 
+  function update(path, withState) {
+    var partialState = getPath(path, appState)
+
+    if (typeof withState === "function") {
+      return update(path, withState(partialState))
+    }
+    if (
+      withState &&
+      (withState = setPath(path, merge(partialState, withState), appState))
+    ) {
+      requestRender((appState = withState))
+    }
+    return appState
+  }
+
+  function getPath(path, source) {
+    return path.length === 0 ? source : getPath(path.slice(1), source[path[0]])
```

Comment:
@JorgeBucaran I think there's a bug here, or at least an inconsistency. Based on this code in `setPath` below(line 100):
```
source != null && name in source
  ? source[name]
  : path[1] >= 0 ? [] : {}
```
I would expect an update for a `state` fragment that doesn't exist yet to work, however this line will throw an error. One possible fix would be to write:
```
return path.length === 0 ? source : getPath(path.slice(1), source[path[0]] || {})
```

instead. I came across the issue because the condition in `setPath` is currently not covered by the tests and I was trying to add coverage.